it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala

```diff
@@ -0,0 +1,239 @@
+    EmbeddedKafka.createCustomTopic("valueOnly")
+    EmbeddedKafka.createCustomTopic("keyAndValue")
+    EmbeddedKafka.createCustomTopic("partitioned", partitions = 5)
+
+    EmbeddedKafka.withProducer[String, String, Unit] { producer =>
+      producer.send(new ProducerRecord("keyOnly", "false", null))
+
+      producer.send(new ProducerRecord("valueOnly", s"{ ${q("key")}: ${q("value")} }"))
+      producer.send(new ProducerRecord("valueOnly", "[1, 2, 3]"))
+      producer.send(new ProducerRecord("valueOnly", q("string")))
+
+      producer.send(new ProducerRecord("keyAndValue", q("key"), q("value")))
+      producer.send(new ProducerRecord("keyAndValue", "[1, 2, 3]", "true"))
+
+      for (number <- 1 to 50) producer.send(new ProducerRecord("partitioned", s"""{ "number": $number }"""))
+    }
+  }
+
+  "Datasource" >> {
+    def baseConfig = Json(
+      "bootstrapServers" := List(s"localhost:${embeddedK.config.kafkaPort}"),
+      "groupId" := "precog",
+      "format" := Json(
+        "type" := "json",
+        "variant" := "line-delimited",
+        "precise" := false)
+    )
+
+    "reads value only topics" >> {
+      def config =
+        ("topics" := List("valueOnly", "partitioned")) ->:
+        ("decoder" := Decoder.rawValue.asJson) ->:
+        baseConfig
+
+      evaluateTyped(config, "valueOnly").unsafeRunSync() must beLike {
+        case Right(jss) => jss must_=== List(
+          Json("key" := jString("value")),
+          Json.array(jNumber(1), jNumber(2), jNumber(3)),
+          jString("string"))
+      }
+    }
+
+    "reads topic keys" >> {
+      def config =
+        ("topics" := List("keyAndValue")) ->:
+          ("decoder" := Decoder.rawKey.asJson) ->:
+          baseConfig
+
+      evaluateTyped(config, "keyAndValue").unsafeRunSync() must beLike {
+        case Right(jss) => jss must_=== List(jString("key"), Json.array(jNumber(1), jNumber(2), jNumber(3)))
+      }
+    }
+
+    "reads topic values" >> {
+      def config =
+        ("topics" := List("keyAndValue")) ->:
+          ("decoder" := Decoder.rawValue.asJson) ->:
+          baseConfig
+
+      evaluateTyped(config, "keyAndValue").unsafeRunSync() must beLike {
+        case Right(jss) => jss must_=== List(jString("value"), jTrue)
+      }
+    }
+
+    "reads keys as empty when keys are absent" >> {
+      def config =
+        ("topics" := List("valueOnly")) ->:
+          ("decoder" := Decoder.rawKey.asJson) ->:
+          baseConfig
+
+      evaluateTyped(config, "valueOnly").unsafeRunSync() must beLike {
+        case Right(jss) => jss must beEmpty
+      }
+    }
+
+    "reads values as empty when values are absent" >> {
+      def config =
+        ("topics" := List("keyOnly")) ->:
+          ("decoder" := Decoder.rawValue.asJson) ->:
+          baseConfig
+
+      evaluateTyped(config, "keyOnly").unsafeRunSync() must beLike {
+        case Right(jss) => jss must beEmpty
+      }
+    }
+
+    "reads partitioned topics" >> {
```

Comment:
other test ideas for partitioned topics:

- a test where some, but not all, of the partitions are empty
- a test where all of the partitions are empty

Reply:
I'm making `keyAndValue` have 5 partitions, so its tests will cover the first case. The second case is covered by the "empty topics" -- it's only one partition, but that makes no difference. There's no special case for "one partition".